test/performance/odp_ipsecfwd.c

```diff
@@ -0,0 +1,1307 @@
+			printf(exposed_algs[i].type == COMBINED_TYPE ? " (combined)" : "");
+
+			odp_ipsec_cipher_capability_t capa[c_cnt];
+
+			(void)odp_ipsec_cipher_capability(exposed_algs[i].idx, capa, c_cnt);
+
+			for (int j = 0; j < c_cnt; ++j)
+				printf(j == 0 ? " (key lengths: %u" : ", %u", capa[j].key_len);
+
+			printf(")\n");
+		}
+	}
+
+	printf("                          Authentication algorithms:\n");
+
+	for (size_t i = 0U; i < len; ++i) {
+		if (exposed_algs[i].type == AUTH_TYPE &&
+		    (ipsec_capa->auths.all_bits & (1 << exposed_algs[i].idx)) > 0U) {
+			a_cnt = odp_ipsec_auth_capability(exposed_algs[i].idx, NULL, 0);
+
+			if (a_cnt < 0)
+				continue;
+
+			printf("                              %d: %s",
+			       exposed_algs[i].idx, exposed_algs[i].name);
+
+			odp_ipsec_auth_capability_t capa[a_cnt];
+
+			(void)odp_ipsec_auth_capability(exposed_algs[i].idx, capa, a_cnt);
+
+			for (int j = 0; j < a_cnt; ++j)
+				printf(j == 0 ? " (key/icv lengths: %u/%u" : ", %u/%u",
+				       capa[j].key_len, capa[j].icv_len);
+
+			printf(")\n");
+		}
+	}
+}
+
+static void print_usage(void)
+{
+	odp_pool_capability_t pool_capa;
+	odp_ipsec_capability_t ipsec_capa;
+
+	if (odp_pool_capability(&pool_capa) < 0) {
+		ODPH_ERR("Error querying pool capabilities\n");
+		return;
+	}
+
+	if (odp_ipsec_capability(&ipsec_capa) < 0) {
+		ODPH_ERR("Error querying IPsec capabilities\n");
+		return;
+	}
+
+	printf("\n"
+	       "Simple IPsec performance tester. Forward and process plain and ipsec packets.\n"
+	       "\n"
+	       "Examples:\n"
+	       "    %s -i ens9f1 -s /etc/odp/sa.conf -f /etc/odp/fwd.conf\n"
+	       "\n"
+	       "    With sa.conf containing, for example:\n"
+	       "        0 222 192.168.1.10 192.168.1.16 4 jWnZr4t7w!zwC*F- 0 2"
+	       " n2r5u7x!A%%D*G-KaPdSg 0 12\n"
+	       "\n"
+	       "    With fwd.conf containing, for example:\n"
+	       "        192.168.1.0 24 ens9f1 aa:bb:cc:dd:11:22\n"
+	       "\n"
+	       "Usage: %s [options]\n"
+	       "\n"
+	       "  -i, --interfaces    Ethernet interfaces for packet I/O, comma-separated,\n"
+	       "                      no spaces.\n"
+	       "  -n, --num_pkts      Number of packet buffers allocated for packet I/O pool.\n"
+	       "                      %u by default.\n"
+	       "  -l, --pkt_len       Maximum size of packet buffers in packet I/O pool. %u by\n"
+	       "                      default.\n"
+	       "  -c, --count         CPU count, %u by default.\n"
+	       "  -m, --mode          Packet output mode.\n"
+	       "                          0: queued output (default)\n"
+	       "                          1: direct output\n"
+	       "  -s, --sa            SA configuration file. Individual SA configuration is\n"
+	       "                      expected to be within a single line, values whitespace\n"
+	       "                      separated:\n"
+	       "\n"
+	       "                          <line in file> Dir SPI SrcIPv4 DstIPv4 CipherAlgoIdx"
```

Comment:
Maybe TunSrcIPv5 and TunDstIPv4 to make it clear that these are IPsec tunnel endpoint addresses?

Reply:
Sure, why not, in v4.